internal/oid/algorithm.go

```diff
@@ -0,0 +1,48 @@
+// Copyright The Notary Project Authors.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package oid
+
+import (
+	"crypto/x509"
+	"encoding/asn1"
+)
+
+// ToSignatureAlgorithm converts ASN.1 digest and signature algorithm
+// identifiers to golang signature algorithms.
+func ToSignatureAlgorithm(digestAlg, sigAlg asn1.ObjectIdentifier) x509.SignatureAlgorithm {
+	switch {
+	case RSA.Equal(sigAlg):
```

Comment:
For each SignerInfo, we need to know the signature algorithm for verification. Therefore, we need both `digestAlg` and `sigAlg` to identify the original signature algorithm.

> Why do we have two ways to obtain the signature algorithm?

I'm not sure what you mean by two ways. Are you suggesting that we can also get the signature algorithm from the certificate, similar to what is done in notation-core-go? I believe we cannot do that for the following reasons:

1. Notation-core-go retrieves the signature algorithm from the certificate based on a fixed mapping. However, this approach is specific to the Notation's definition. We assume that a key with RSA 2048 should always sign with RSASSA-PSS with SHA-256, but in reality, an RSA 2048 key can also sign using RSASSA-PSS with SHA-512. Therefore, we cannot apply the same logic to the CMS package.
2. The `certificates` field is optional, so we should not rely on an optional field to obtain the signature algorithm.
3. The `digestAlgorithm` (`DigestAlgorithmIdentifier`) and `signatureAlgorithm` (`SignatureAlgorithmIdentifier`) in the `SignerInfo` are specifically designed fields in RFC 5652 to determine the signature algorithm, so we should use them directly.

Reply:
Update: I may have misunderstood your question. Let me explain it again:

https://www.rfc-editor.org/rfc/rfc5652#section-10.1.2
>  The SignatureAlgorithmIdentifier type identifies a signature
   algorithm, and it can also identify a message digest algorithm.

Therefore, it is necessary to account for the possibility that the signature algorithm may or may not include the digest algorithm.